KAFKA-8592; Fix for resolving variables for dynamic config as per KIP-421.

jeffhuang26 opened the pull request:
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

jeffhuang26 committed:
Fix for resolving variables for dynamic config as per KIP-421.

jeffhuang26 commented:
This PR added a fix on top of following PR and other codes from Tejal. I created a new PR due to start from clean branch.
https://github.com/apache/kafka/pull/6992

stanislavkozlovski commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -371,11 +372,13 @@ class AdminManager(val config: KafkaConfig,
+          configPropsOrigin.setProperty(configEntry.name, configEntry.value)
         }
 
+        val configProps = resolveVariableConfigs(configPropsOrigin)
```
Do we also want to include this resolving logic in the `IncrementalAlterConfigs` API too?

jeffhuang26 replied:
I reverted changes because this solution will write resolved config(clear text), which is not right,  to ZK. I moved resolveVariableConfigs to DynamicBrokerConfig. So this will cover IncrementalAlterConfigs scenario that will call alterBrokerConfigs. See latest commit.

stanislavkozlovski commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -607,4 +610,17 @@ class AdminManager(val config: KafkaConfig,
+
+  private def resolveVariableConfigs(propsOriginal: Properties): Properties = {
+    val props = new Properties
+    propsOriginal.put("bootstrap.servers", "localhost:1")
```
What is the purpose of this `put`? As far as I can tell, it won't get added to the `props` object anyway, as we filter on the key below

jeffhuang26 replied:
I reverted changes because this solution will write resolved config(clear text), which is not right, to ZK. See latest commit.

stanislavkozlovski commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -607,4 +610,17 @@ class AdminManager(val config: KafkaConfig,
+  private def resolveVariableConfigs(propsOriginal: Properties): Properties = {
+    val props = new Properties
+    propsOriginal.put("bootstrap.servers", "localhost:1")
+    val config = new AdminClientConfig(propsOriginal)
```
Given that we don't use `AdminClientConfig`'s overridden `postProcessParsedConfig()` method, does it make sense to use `AbstractConfig` here?

jeffhuang26 replied:
Updated code using AbstractConfig. See latest commit.

stanislavkozlovski commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -337,7 +367,7 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
 
     // Update internal keystore/truststore and validate new client connections from broker (e.g. controller).
-    // Alter internal keystore from `sslProperties1` to `sslProperties2`, force disconnect of a controller connection
+    // Alter internal keystore testKeyStoreAlter_ConfigProviderfrom `sslProperties1` to `sslProperties2`, force disconnect of a controller connection
```
nit: typo

jeffhuang26 replied:
Fixed typo.

jeffhuang26 committed:
Merge branch 'trunk' of https://github.com/apache/kafka into trunk

jeffhuang26 committed:
Updated code based on comments, fixed bug that write resolved config(clear text) to zk.

jeffhuang26 committed:
Removed unused code.

rajinisivaram commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -216,6 +216,34 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
   }
 
+  @Test
+  def testKeyStoreAlter_ConfigProvider(): Unit = {
```
We should rename this test `testUpdatesUsingConfigProvider` and keep all testing for config providers within this class. DynamicBrokerReconfigurationTest and the underlying updates are very complex and adding config providers in existing tests makes it even more difficult to debug. Having all config provider tests in one method would isolate provider related issues from other config update issues.

jeffhuang26 replied:
I renamed the new testing function to testUpdatesUsingConfigProvider and put all testing cases to the function.

jeffhuang26 replied:
I removed the function. Reuse function verifyState.

rajinisivaram commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -230,6 +258,7 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
     val producer1 = ProducerBuilder().trustStoreProps(sslProperties2).maxRetries(0).build()
     verifyAuthenticationFailure(producer1)
 
+    sslProperties2.setProperty("ssl.keystore.password", "${file:/usr/kerberos:password}")
```
as mentioned in the previous comment, we should revert this.

jeffhuang26 replied:
reverted it to original version.

rajinisivaram commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -363,7 +393,9 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
 
     val props = new Properties
-    props.put(KafkaConfig.LogCleanerThreadsProp, "2")
+    props.put(KafkaConfig.LogCleanerThreadsProp, "${file:/usr/kerberos:log}")
```
It is confusing to have a config name `file:/usr/kerberos:log` in a test that is doesn't use Kerberos and has a large number of log configs. We should move this to the config providers test method. And change the config name to `log.cleaner.threads`.

jeffhuang26 replied:
I changed it. Not sure what exactly you suggested. Please see new version and give new comment on this.

rajinisivaram commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -363,7 +393,9 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
-    props.put(KafkaConfig.LogCleanerThreadsProp, "2")
+    props.put(KafkaConfig.LogCleanerThreadsProp, "${file:/usr/kerberos:log}")
+    props.setProperty("config.providers", "file")
+    props.setProperty("config.providers.file.class", "org.apache.kafka.common.config.provider.MockFileConfigProvider")
```
`org.apache.kafka.common.config.provider.MockFileConfigProvider` is basically a class with one line of code. It will be better to create a copy of it to use in this test rather than expect a class in another project to provide the config names and values that this test relies on. We can then use meaningful path and config names here.

jeffhuang26 replied:
I recreated a local copy. Please see new code updated. I still keep the org.apache.kafka.common.config.provider.MockFileConfigProvider. Do I need delete it or keep it?

rajinisivaram reviewed (commented):
@jeffhuang26 Thanks for the PR, looks good overall. Left a few comments regarding tests. We should have one integration test under `DynamicBrokerReconfigurationTest` for updates using config provider. They dont need to test the effect of every config change since that is tested elsewhere. They just need to test that configs have been updated - e.g. using describe after the update. We need to validate the following:
1) Describe after update and check that update has happened. We should do this for String, Password and Integer configs as the PR currently does, but do this in one method
2) Do updates using ConfigCommand and make sure that works for the three config types
3) Ensure ZooKeeper contains unresolved config
4) Ensure we dont perform update when a config hasn't changed (this is important because if we compare unresolved config with resolved config to verify if config has changed, there will be a lot of unnecessary expensive reconfiguration)
5) Verify that AdminClient#describeConfigs returns resolved values

jeffhuang26 committed:
Merge branch 'trunk' of https://github.com/apache/kafka into trunk

jeffhuang26 committed:
Updated code based on review comments.

jeffhuang26 committed:
Reverted change to excactly original version.

jeffhuang26 commented:
> @jeffhuang26 Thanks for the PR, looks good overall. Left a few comments regarding tests. We should have one integration test under `DynamicBrokerReconfigurationTest` for updates using config provider. They dont need to test the effect of every config change since that is tested elsewhere. They just need to test that configs have been updated - e.g. using describe after the update. We need to validate the following:
> 
> 1. Describe after update and check that update has happened. We should do this for String, Password and Integer configs as the PR currently does, but do this in one method
> 2. Do updates using ConfigCommand and make sure that works for the three config types
> 3. Ensure ZooKeeper contains unresolved config
> 4. Ensure we dont perform update when a config hasn't changed (this is important because if we compare unresolved config with resolved config to verify if config has changed, there will be a lot of unnecessary expensive reconfiguration)
> 5. Verify that AdminClient#describeConfigs returns resolved values

I think I did what you suggested except two thing. 
1. The ssl store key password is always null using AdminClient#describeConfigs even though it can be verified by both waitForConfig after executing ConfigCommand and result fetched back from ZK. I can see the password is stored in config.originals, but it will not be shown when it is sent back to ConfigCommand over api. I commented out two lines for this for now.
2. I can verify update action will not be performed if I update same config with config using provider by manually traces that show "updateBrokerConfig in DynamicBrokerConfig will not called". I do not know how to verify it in other way.
Notes: 
a. Program will treat config value with and without config provider different for same config name. That mean program will perform update if value changed from non-config provider to config provider and vice versa.
b.  Program will not perform update if value is same for same config.

rajinisivaram commented:
@jeffhuang26 Thanks for the updates.
1) You are right, we don't allow passwords to be retrieved using admin client. It should be sufficient to verify the password in server.config.originals for Password configs.
2) One option for verifying if update has taken place is to use a custom metrics reporter that counts reconfigurations. For an example of reconfigurable metrics reporter, you can take a look at `DynamicBrokerReconfigurationTest.testMetricsReporterUpdate`. You could do something similar in `testUpdatesUsingConfigProvider`.

jeffhuang26 committed:
Added testing case for veryfing updating for property using config provider.

jeffhuang26 committed:
Merge branch 'trunk' of https://github.com/apache/kafka into trunk

jeffhuang26 commented:
> @jeffhuang26 Thanks for the updates.
> 
> 1. You are right, we don't allow passwords to be retrieved using admin client. It should be sufficient to verify the password in server.config.originals for Password configs.
> 2. One option for verifying if update has taken place is to use a custom metrics reporter that counts reconfigurations. For an example of reconfigurable metrics reporter, you can take a look at `DynamicBrokerReconfigurationTest.testMetricsReporterUpdate`. You could do something similar in `testUpdatesUsingConfigProvider`.

I updated code using TestMetricsReporter for testing updating scenario. I also fixed a few of places in existing code, please double check fixes.
1. fixed one bug in function verifyState in class TestMetricsReporter that replaced "assertEquals(reconfigureCount, reconfigureCount)" with assertEquals(reconfigureCount, this.reconfigureCount) because "assertEquals(reconfigureCount, reconfigureCount)" will always be true.
2. fixed a  few of places in function testMetricsReporterUpdate caused by above fixing.

stanislavkozlovski commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -288,11 +298,23 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
+      reporter.verifyStateUpdUsingConfigProvider(reconfigureCount = 0, deleteCount = 0, logCleanerThreads = 2)
+    }
+
+    // 2. verify udpate occure if the value of property changing.
```
nit: udpate typo

jeffhuang26 replied:
thanks.

stanislavkozlovski commented on clients/src/test/java/org/apache/kafka/common/config/provider/MockFileConfigProvider.java:
```diff
@@ -24,6 +24,6 @@
     @Override
     protected Reader reader(String path) throws IOException {
-        return new StringReader("key=testKey\npassword=randomPassword");
+        return new StringReader("key=testKey\npassword=ServerPassword\na=b\nlog=2");
```
Do we still need this? I see we use our own `kafka.server.MockFileConfigProvider` in the tests

jeffhuang26 replied:
We used this class for a few of testing cases in testing AbstractConfigTest. I think we might also keep this one for future testing. What you think? @rajinisivaram

rajinisivaram replied:
We can keep the class since it was being used earlier. But we don't need to change it in this PR right?

jeffhuang26 replied:
Reverted change.

rajinisivaram commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -177,6 +177,18 @@ object DynamicBrokerConfig {
+    val config = new AbstractConfig(new ConfigDef(), propsOriginal, false)
+    val resolvedProps = config.originals
+    for ((key, value) <- resolvedProps.asScala) {
+      if (!key.startsWith("config.provider")) {
```
Shouldn't this be `config.providers`? We could use AbstractConfig.CONFIG_PROVIDERS_PARAM here?

jeffhuang26 replied:
Use config.providers is more accurate than config.provider, but result is same. I updated it to config.providers.

rajinisivaram commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -177,6 +177,18 @@ object DynamicBrokerConfig {
+      if (!key.startsWith("config.provider")) {
+        props.put(key, value)
+      }
+    }
```
The loop can be replaced with:
```
props.putAll(config.originals.asScala.filter(_._1.startsWith(AbstractConfig.CONFIG_PROVIDERS_CONFIG)).asJava)
```

jeffhuang26 replied:
Updated code using your suggestion. But I keep "config.providers" because CONFIG_PROVIDERS_CONFIG is private in AbstractConfig.

jeffhuang26 replied:
Updated the AbstractConfig to make the CONFIG_PROVIDERS_CONFIG public.

rajinisivaram commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -1646,16 +1774,33 @@ class TestMetricsReporter extends MetricsReporter with Reconfigurable with Close
     assertEquals(pollingInterval, this.pollingInterval)
   }
 
+  def verifyStateUpdUsingConfigProvider(reconfigureCount: Int, deleteCount: Int, logCleanerThreads: Int): Unit = {
```
Perhaps `verifyUpdatedStateUsingConfigProvider` would be better since `Upd` seems a bit cryptic?

jeffhuang26 replied:
Updated it.

rajinisivaram commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -1599,6 +1707,9 @@ class TestMetricsReporter extends MetricsReporter with Reconfigurable with Close
   @volatile var clusterUpdateCount = 0
   @volatile var pollingInterval: Int = -1
+  @volatile var isPollingInterval: Boolean = false
+  @volatile var logCleanerThreads: Int = -1
```
If we want to reuse this class, we should just use the existing `polling.interval` config and update that rather than `log.cleaner.threads`. Log cleaner threads looks really odd in a metrics reporter. We can then remove most of the changes from this class.

jeffhuang26 replied:
Good suggestion, I can reuse the polling.interval with config provider. I will update other changes too.

jeffhuang26 replied:
Updated code to use polling.interval for Integer type testing.

rajinisivaram commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -216,6 +217,106 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
+    val configPrefix = listenerPrefix(SecureExternal)
+    var brokerConfigs = describeConfig(adminClients.head, servers).entries.asScala
+    // the following are values before updated
+    assertTrue("Initial value of log cleaner threads", (brokerConfigs.find(_.name == KafkaConfig.LogCleanerThreadsProp).get).value == "1")
```
unnecessary parenthesis: there are a few of these in this method

rajinisivaram commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -216,6 +217,106 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
+    // setup ssl properties
+    val secProps = securityProps(sslProperties1, KEYSTORE_PROPS, configPrefix)
+
+    // configure configure config providers and properties need be updated
```
just `configure` (once) in the comment

jeffhuang26 replied:
Updated it.

rajinisivaram commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -216,6 +217,106 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
+    val updatedProps = new Properties
+    updatedProps.setProperty("config.providers", "file")
+    updatedProps.setProperty("config.providers.file.class", "kafka.server.MockFileConfigProvider")
+    // configure MetricReport
```
typo in comment, comment unnecessary anyway

rajinisivaram commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -216,6 +217,106 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
+
+    // describe using AdminClient. values should be resolved
+    brokerConfigs = describeConfig(adminClients.head, servers).entries.asScala
+    assertTrue("log cleaner thread is not updated", (brokerConfigs.find(_.name == KafkaConfig.LogCleanerThreadsProp).get).value == "2")
```
unnecessary parenthesis

jeffhuang26 replied:
Updated it,

rajinisivaram commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -216,6 +217,106 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
+    assertTrue("store type is not updated", (brokerConfigs.find(_.name == configPrefix+KafkaConfig.SslTruststoreTypeProp).get).value == "JKS")
+
+    // verify the update
+    // 1. verify update not occuring if the value of property is same using same updatedProps.
```
typo: occurring

rajinisivaram commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -216,6 +217,106 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
 
+  @Test
+  def testUpdatesUsingConfigProvider(): Unit = {
+    val LOG_CLEANER_THREADS_VAL = "${file:log.cleaner.threads:log}"
```
Don't think we do caps for these vals in scala.

jeffhuang26 replied:
Updated it.

rajinisivaram reviewed (commented):
@jeffhuang26 Thanks for the updates, looks good. Left some minor comments.

jeffhuang26 committed:
Merge branch 'trunk' of https://github.com/apache/kafka into trunk

jeffhuang26 committed:
Merge branch 'trunk' of https://github.com/apache/kafka into trunk

jeffhuang26 committed:
Updated code based on review comments.

jeffhuang26 commented:
Hi Rajini and Stanislav,

Any comments on last update?

Jeff.

rajinisivaram commented:
retest this please

rajinisivaram reviewed (approved):
@jeffhuang26 Thanks for the updates, LGTM. Running tests again since results from the old run are not available to see if they are known transient failures.

jeffhuang26 commented:
> retest this please

Will do.

rajinisivaram commented:
@jeffhuang26 There is a compilation failure in the build here: https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/6907/console

stanislavkozlovski commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -216,6 +217,98 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
+      }
+
+      servers.foreach { server =>
+        val args = Array("--bootstrap-server", TestUtils.bootstrapServers(servers, new ListenerName(SecureInternal)),
```
nit: we have this logic duplicated in `private def alterSslKeystoreUsingConfigCommand`, perhaps a helper method could help

jeffhuang26 committed:
Merge branch 'trunk' of https://github.com/apache/kafka into trunk

jeffhuang26 committed:
Fixed compile errors on Java 11.

jeffhuang26 commented:
I fixed compile errors on JDK 11.

jeffhuang26 commented:
> > retest this please
> 
> Will do.

Fixed compile error on JDK 11.

jeffhuang26 commented:
> @jeffhuang26 There is a compilation failure in the build here: https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/6907/console

putAll() function got compiled error on JDK 11. I updated code not use putAll.

jeffhuang26 committed:
Merge branch 'trunk' of https://github.com/apache/kafka into trunk

jeffhuang26 committed:
Updated code based on review comments.

rajinisivaram commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -216,6 +217,93 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
+      }
+
+      servers.foreach { server =>
+        val args = buildConfigCommandParameters(server, propsFile.getAbsolutePath, props)
```
Hmm. it would be better to create a method `alterConfigsUsingConfigCommand` that does almost everything in this method which was copied from `alterSslKeystoreUsingConfigCommand`.

```
private def alterSslKeystoreUsingConfigCommand(props: Properties, listener: String): Unit = {
    val configPrefix = listenerPrefix(listener)
    val newProps = securityProps(props, KEYSTORE_PROPS, configPrefix)
    alterConfigsUsingConfigCommand(newProps)
   waitForConfig(s"$configPrefix$SSL_KEYSTORE_LOCATION_CONFIG", props.getProperty(SSL_KEYSTORE_LOCATION_CONFIG))
}
```
`executeConfigCommand` can be removed and in its place, you would just call
```
alterConfigsUsingConfigCommand(props)
```

jeffhuang26 replied:
Actually it is much better. Thanks for finding this.

jeffhuang26 committed:
Updated code based on review.

rajinisivaram reviewed (approved):
@jeffhuang26 Thanks for the updates, LGTM. Merging to trunk.
